### PR TITLE
refactor(XMLHttpRequest): Use implementation and generated wrapper

### PR DESCRIPTION
**Pull request by ExE-Boss**
Part&nbsp;of&nbsp;#2727.

---

Depends&nbsp;on&nbsp;#2764, please&nbsp;merge&nbsp;that&nbsp;first.

Best&nbsp;reviewed&nbsp;with whitespace‑only changes&nbsp;ignored: [#2737/files?w=1](https://github.com/jsdom/jsdom/pull/2737/files?w=1).

---

review?(@domenic, @TimothyGu): I&nbsp;managed to&nbsp;get the&nbsp;tests to&nbsp;pass.

### CODE REVIEW

**Comment by domenic** [comment]
I just want to say thanks so much for working on this. We might not be able to review it too quickly, given all the other activity going on around multi-globals and such. And there will probably be some rebasing to be done, e.g. in #2743 I am changing some of the XHR code to use isArrayBuffer instead of instanceof. But this will be really great to have.

**Comment by ExE-Boss** [comment]
<https://github.com/jsdom/jsdom/pull/2743>&nbsp;has&nbsp;been&nbsp;merged, so&nbsp;this&nbsp;should now&nbdp;be&nbsp;unblocked.

**Comment by ExE-Boss** [comment]
review?(@domenic) I’ve&nbsp;now&nbsp;resolved the&nbsp;merge&nbsp;conflicts introduced&nbsp;by&nbsp;<https://github.com/jsdom/jsdom/pull/2733>.

Note&nbsp;that we&nbsp;should get&nbsp;<https://github.com/jsdom/jsdom/pull/2764>&nbsp;merged before&nbsp;this, since&nbsp;this&nbsp;will&nbsp;otherwise break&nbsp;<code>git&nbsp;blame</code>.

**Comment by domenic** [comment]
Can you remove the jsdoc comments? That will reduce the diff substantially and make review easier.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 1]
```diff
@@ -0,0 +1,1109 @@
...
+
+  /**
+   * @param {string} header
+   * @param {string} value
+   */
+  setRequestHeader(header, value) {
+    const { flag, properties } = this;
+
+    header = conversions.ByteString(header);
+    value = conversions.ByteString(value);
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: These two should be done by the webidl2js layer.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 2]
```diff
@@ -0,0 +1,1109 @@
...
+    if (this.readyState === XMLHttpRequest.LOADING || this.readyState === XMLHttpRequest.DONE) {
+      throw DOMException.create(this._globalObject, ["The object is in an invalid state.", "InvalidStateError"]);
+    }
+    if (this.readyState === XMLHttpRequest.OPENED && flag.synchronous) {
+      throw DOMException.create(this._globalObject, [
+        "The object does not support the operation or argument.",
+        "InvalidAccessError"
+      ]);
+    }
+    if (!XMLHttpRequestResponseType.has(responseType)) {
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: This should be taken care of by webidl2js.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+
+class XMLHttpRequest extends XMLHttpRequestEventTargetImpl {
+  constructor(window) {
+    super(window);
+
+    /** @type {import("../../browser/Window")} */
+    this._globalObject = window;
+    /** @type {import("./XMLHttpRequestUpload-impl.js").implementation} */
+    this.upload = XMLHttpRequestUpload.createImpl(window);
+    this.upload._ownerDocument = this._ownerDocument;
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: I don't see this being used anywhere that's obvious… is this necessary? In either case it seems to belong more in `XMLHttpRequestUpload`'s impl constructor.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+    }
+  }
+  getAllResponseHeaders() {
+    const { properties, readyState } = this;
+    if (readyState === XMLHttpRequest.UNSENT || readyState === XMLHttpRequest.OPENED) {
+      return "";
+    }
+    return Object.keys(properties.responseHeaders)
+      .filter(key => properties.filteredResponseHeaders.indexOf(key) === -1)
+      .map(key => [conversions.ByteString(key).toLowerCase(), properties.responseHeaders[key]].join(": "))
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: The extra `ByteString` conversion doesn't seem necessary.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 5]
```diff
@@ -0,0 +1,1109 @@
...
+    }
+
+    flag.uri = urlObj.href;
+    flag.requestHeaders = {};
+    flag.preflight = false;
+
+    properties.send = false;
+    properties.uploadListener = false;
+    properties.abortError = false;
+    this.responseURL = "";
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: The original also resets `properties.requestBuffer` and `properties.requestCache`. Is that unnecessary?

**Comment by TimothyGu** [comment]
A rebase looks needed.

And also, yes, please remove the JSDoc comments. They are fine during development for IDE support, but we don't accept them in the code base itself.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 6]
```diff
@@ -83,945 +81,881 @@ const corsSafeResponseHeaders = new Set([
...
+setupForSimpleEventAccessors(XMLHttpRequestImpl.prototype, ["readystatechange"]);
 
-  addConstants(XMLHttpRequest, {
-    UNSENT: 0,
-    OPENED: 1,
-    HEADERS_RECEIVED: 2,
-    LOADING: 3,
-    DONE: 4
-  });
+addConstants(XMLHttpRequestImpl, {
```
Comment (domenic) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]:
```suggestion
Object.assign(XMLHttpRequestImpl, {
```

We should eventually get rid of `addConstants`, so maybe it'd be good to start here. (Since webidl2js will now take care of the property descriptors.)

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 7]
```diff
@@ -83,945 +81,881 @@ const corsSafeResponseHeaders = new Set([
...
-      }
-
-      this[xhrSymbols.flag].overrideMIMEType = "application/octet-stream";
+    this.flag.overrideMIMEType = "application/octet-stream";
 
-      // Waiting for better spec: https://github.com/whatwg/xhr/issues/157
-      const parsed = MIMEType.parse(mime);
-      if (parsed) {
-        this[xhrSymbols.flag].overrideMIMEType = parsed.essence;
+    // Waiting for better spec: https://github.com/whatwg/xhr/issues/157
```
Comment (domenic) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: Optional: this spec issue is closed so we could update this to match the spec.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 8]
```diff
@@ -83,945 +81,881 @@ const corsSafeResponseHeaders = new Set([
...
-              if (charset && !asciiCaseInsensitiveMatch(charset, encoding) && encoding !== null) {
-                parsed.parameters.set("charset", encoding);
-                xhrUtils.updateRequestHeader(flag.requestHeaders, "content-type", parsed.toString());
-              }
+        const existingContentType = xhrUtils.getRequestHeader(flag.requestHeaders, "content-type");
+        if (mimeType !== null && existingContentType === null) {
+          flag.requestHeaders["Content-Type"] = mimeType;
+        } else if (existingContentType !== null && encoding !== null) {
+          // Waiting for better spec: https://github.com/whatwg/xhr/issues/188. This seems like a good guess at what
+          // the spec will be, in the meantime.
```
Comment (domenic) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: Optional: this spec issue is closed so we could update this to match the spec.

File: lib/jsdom/living/xhr/xhr-utils.js [hunk 9]
```diff
@@ -126,18 +122,19 @@ function requestErrorSteps(xhr, event, exception) {
...
+  xhr.status = 0;
+  xhr.statusText = "";
 }
 
-// return a "request" client object or an event emitter matching the same behaviour for unsupported protocols
-// the callback should be called with a "request" response object or an event emitter matching the same behaviour too
+/**
+ * return a "request" client object or an event emitter matching the same behaviour for unsupported protocols
+ * the callback should be called with a "request" response object or an event emitter matching the same behaviour too
+ */
```
Comment (domenic) [inline lib/jsdom/living/xhr/xhr-utils.js]: Can you undo the change to the comment syntax?

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 10]
```diff
@@ -83,945 +81,881 @@ const corsSafeResponseHeaders = new Set([
...
+      filteredResponseHeaders: [],
+
+      error: "",
+      uploadComplete: false,
+      uploadListener: false,
+
+      /**
+       * Signifies that we're calling abort() from xhr-utils.js because of a window shutdown.
+       * In that case the termination reason is "fatal", not "end-user abort".
+       */
```
Comment (domenic) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: Can you undo the change to the comment syntax?

**Comment by domenic** [comment]
@ExE-Boss any ideas on why the tests are failing now? I think they were not failing a few commits ago, right?

**Comment by ExE-Boss** [comment]
Well, the&nbsp;only&nbsp;thing that&nbsp;changed as&nbsp;a&nbsp;result of&nbsp;replacing&nbsp;`addConstants(…)` with&nbsp;property&nbsp;assignment is&nbsp;that the&nbsp;constants were&nbsp;no&nbsp;longer on&nbsp;the&nbsp;`XMLHttpRequestImpl`&nbsp;prototype.

**Comment by ExE-Boss** [comment]
@domenic The&nbsp;tests&nbsp;are&nbsp;passing now&nbsp;that&nbsp;I’ve&nbsp;added the&nbsp;constants&nbsp;back to&nbsp;`XMLHttpRequestImpl.prototype`, so&nbsp;there’s probably&nbsp;something in&nbsp;**JSDOM** that&nbsp;gets&nbsp;them from&nbsp;the&nbsp;instance’s&nbsp;implementation rather&nbsp;than the&nbsp;constructor.

**Comment by domenic** [comment]
@TimothyGu can you take a final look? I think after this PR I'll do the new major version...

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 11]
```diff
@@ -83,945 +80,883 @@ const corsSafeResponseHeaders = new Set([
...
+      // In that case the termination reason is "fatal", not "end-user abort".
+      abortError: false,
+
+      cookieJar: _ownerDocument._cookieJar,
+      bufferStepSize: 1 * 1024 * 1024, // pre-allocate buffer increase step size. init value is 1MB
+      totalReceivedChunkSize: 0
+    };
+
+    Object.setPrototypeOf(this.flag, null);
+    Object.setPrototypeOf(this.properties, null);
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: I don't think these are necessary.

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 12]
```diff
@@ -83,945 +80,883 @@ const corsSafeResponseHeaders = new Set([
...
-    configurable: true
-  });
+setupForSimpleEventAccessors(XMLHttpRequestImpl.prototype, ["readystatechange"]);
 
-  setupForSimpleEventAccessors(XMLHttpRequest.prototype, ["readystatechange"]);
+Object.defineProperty(XMLHttpRequestImpl, "UNSENT", { value: 0, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl, "OPENED", { value: 1, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl, "HEADERS_RECEIVED", { value: 2, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl, "LOADING", { value: 3, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl, "DONE", { value: 4, enumerable: true });
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: Can we just use simple assignment? I don't think the writability/configurability settings have too many benefits (as in I don't think anyone would do `XMLHttpRequestImpl.UNSENT = …` in their code).

```js
// Make constants available on both the class and its instances, like in Web IDL.
for (const obj of [XMLHttpRequestImpl, XMLHttpRequestImpl.prototype]) {
  obj.UNSENT = 0;
  // …
}
```

File: lib/jsdom/living/xhr/XMLHttpRequestEventTarget-impl.js [hunk 13]
```diff
@@ -1,10 +1,15 @@
...
 const EventTargetImpl = require("../events/EventTarget-impl").implementation;
+const idlUtils = require("../generated/utils");
 const { setupForSimpleEventAccessors } = require("../helpers/create-event-accessor");
 
 const events = ["loadstart", "progress", "abort", "error", "load", "timeout", "loadend"];
 
-class XMLHttpRequestEventTargetImpl extends EventTargetImpl {}
+class XMLHttpRequestEventTargetImpl extends EventTargetImpl {
+  get _ownerDocument() {
+    return idlUtils.implForWrapper(this._globalObject._document);
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequestEventTarget-impl.js]: Can this ever change? If not (and I think not) then I'd prefer assigning it statically in the constructor for `XMLHttpRequestEventTargetImpl`.

More generally I feel this is a vestige of the days where we didn't have `._globalObject` available on every IDL object. But now that we do, maybe it's time for `EventTargetImpl` to use `this._globalObject._document` directly? Maybe something to think about later. @domenic @pmdartus

File: lib/jsdom/living/xhr/XMLHttpRequest-impl.js [hunk 14]
```diff
@@ -83,945 +80,883 @@ const corsSafeResponseHeaders = new Set([
...
-    OPENED: 1,
-    HEADERS_RECEIVED: 2,
-    LOADING: 3,
-    DONE: 4
-  });
+Object.defineProperty(XMLHttpRequestImpl.prototype, "UNSENT", { value: 0, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl.prototype, "OPENED", { value: 1, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl.prototype, "HEADERS_RECEIVED", { value: 2, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl.prototype, "LOADING", { value: 3, enumerable: true });
+Object.defineProperty(XMLHttpRequestImpl.prototype, "DONE", { value: 4, enumerable: true });
```
Comment (TimothyGu) [inline lib/jsdom/living/xhr/XMLHttpRequest-impl.js]: Looks like the only place in jsdom that gets the constant from the instance is `lib/jsdom/living/xhr/xhr-sync-worker.js` and `lib/jsdom/living/xhr/xhr-utils.js`. Best to fix those for once and for all.

File: lib/jsdom/living/xhr/XMLHttpRequestEventTarget-impl.js (hunk 13 above)
Comment (ExE-Boss) [reply lib/jsdom/living/xhr/XMLHttpRequestEventTarget-impl.js]: The&nbsp;`_document`&nbsp;property gets&nbsp;cleared when&nbsp;`window.close()` is&nbsp;called.

This&nbsp;caused&nbsp;several WPT&nbsp;failures when&nbsp;I&nbsp;tried assigning&nbsp;it&nbsp;statically.

Comment (domenic) [reply lib/jsdom/living/xhr/XMLHttpRequestEventTarget-impl.js]: I do agree there's probably a better way to do this these days, but we can do it in a separate PR. I'll push a TODO comment at least and then merge this.
